src/c2goto/library/stdlib.c

```diff
@@ -8,15 +8,65 @@
 #undef atol
 #undef getenv
 
+typedef struct atexit_key
+{
+  void (*atexit_func)();
+  struct atexit_key *next;
+} __ESBMC_atexit_key;
+
+static __ESBMC_atexit_key *__ESBMC_atexits = NULL;
+
+void __atexit_handler()
+{
+__ESBMC_HIDE:;
+  // This is here to prevent k-induction from unwind the next loop unnecessarily
+  if(__ESBMC_atexits == NULL)
+    return;
+
+  while(__ESBMC_atexits)
+  {
+    __ESBMC_atexits->atexit_func();
+    __ESBMC_atexit_key *__ESBMC_tmp = __ESBMC_atexits->next;
+    free(__ESBMC_atexits);
+    __ESBMC_atexits = __ESBMC_tmp;
+  }
+}
+
+int atexit(void (*func)(void))
+{
+__ESBMC_HIDE:;
+  __ESBMC_atexit_key *l =
+    (__ESBMC_atexit_key *)malloc(sizeof(__ESBMC_atexit_key));
+  if(l == NULL)
+    return -1;
+  l->atexit_func = func;
+  l->next = __ESBMC_atexits;
+  __ESBMC_atexits = l;
+  return 0;
+}
+
 #pragma clang diagnostic push
 #pragma GCC diagnostic ignored "-Winvalid-noreturn"
 void exit(int status)
 {
+__ESBMC_HIDE:;
+  __atexit_handler();
+  __ESBMC_finish_formula();
   __ESBMC_assume(0);
 }
 
 void abort(void)
 {
+__ESBMC_HIDE:;
+  __atexit_handler();
```

Comment:
Not sure about this.  From [Linux Die](https://linux.die.net/man/3/atexit)
```
The atexit() function registers the given function to be called at normal process termination, either via [exit](https://linux.die.net/man/3/exit)(3) or via return from the program's main().
```
I.e., abort should not call the atexit functions.

Did you do this for sv-comp?

Reply:
actually, I'm not sure why I added that... I think we assume abort is a valid program termination in sv-comp? I guess we can remove that and re run the sv-comp and see if we get any regression.